scapy/contrib/automotive/obd/scanner.py

```diff
@@ -0,0 +1,144 @@
+#! /usr/bin/env python
+# flake8: noqa: F405
+
+# This file is part of Scapy
+# See http://www.secdev.org/projects/scapy for more information
+# Copyright (C) Andreas Korb <andreas.korb@e-mundo.de>
+# Copyright (C) Friedrich Feigel <friedrich.feigel@e-mundo.de>
+# This program is published under a GPLv2 license
+
+import sys
+from scapy.compat import chb
+from scapy.contrib.automotive.obd.obd import *
+
+
+def _supported_id_numbers(socket, timeout, service_class, id_name, verbose):
+    # timeout only required for the OBD Simulator, since it might tell it
+    # supports a PID, while it actually doesn't and won't respond to this PID.
+    # If this happens with a real ECU, it is an implementation error.
+
+    supported_id_numbers = set()
+    supported_prop = 'supported_' + id_name + 's'
+
+    # ID 0x00 requests the first range of supported IDs in OBD
+    supported_ids_req = OBD()/service_class(b'\x00')
+
+    while supported_ids_req is not None:
+        resp = socket.sr1(supported_ids_req, timeout=timeout, verbose=verbose)
+
+        # If None, the device did not respond.
+        # Usually only occurs, if device is off.
+        if resp is None:
+            break
+
+        supported_ids_req = None
+
+        all_supported_in_range = getattr(resp.data_records[0], supported_prop)
+
+        for supported in all_supported_in_range:
+            id_number = int(supported[-2:], 16)
+            supported_id_numbers.add(id_number)
+
+            # check whether supported id is for next id range
+            if id_number % 20 == 0:
```

Comment:
this should be 0x20, shouldn't it?

Reply:
Absolutely.
It contains the Bits indicating, whether the next 32 PIDs are supported or not.